### PR TITLE
feat: Add read themes from github repo

### PR DESCRIPTION
**Pull request by neretin-trike**
Сделал возможность считывания и парскинг темы из нашего репозитория. Для доступ к уже существующей необходимо перейти по ссылке вида: https://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-270/?theme=<theme_name>

:baby_chick: Download canary assets:

[colors--canary.270.3956267287.xml](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/colors--canary.270.3956267287.xml)  
[color_metro_ios-swift--canary.270.3956267287.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_ios-swift--canary.270.3956267287.swift)  
[color_metro_kotlin--canary.270.3956267287.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_kotlin--canary.270.3956267287.kt)  
[color_metro_react-native--canary.270.3956267287.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_react-native--canary.270.3956267287.ts)  
[color_sbermarket_ios-swift--canary.270.3956267287.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_ios-swift--canary.270.3956267287.swift)  
[color_sbermarket_kotlin--canary.270.3956267287.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_kotlin--canary.270.3956267287.kt)  
[color_sbermarket_react-native--canary.270.3956267287.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_react-native--canary.270.3956267287.ts)  
[color_sberprime_ios-swift--canary.270.3956267287.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_ios-swift--canary.270.3956267287.swift)  
[color_sberprime_kotlin--canary.270.3956267287.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_kotlin--canary.270.3956267287.kt)  
[color_sberprime_react-native--canary.270.3956267287.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_react-native--canary.270.3956267287.ts)  
[color_selgros_ios-swift--canary.270.3956267287.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_ios-swift--canary.270.3956267287.swift)  
[color_selgros_kotlin--canary.270.3956267287.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_kotlin--canary.270.3956267287.kt)  
[color_selgros_react-native--canary.270.3956267287.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_react-native--canary.270.3956267287.ts)  
[color_smbusiness_ios-swift--canary.270.3956267287.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_ios-swift--canary.270.3956267287.swift)  
[color_smbusiness_kotlin--canary.270.3956267287.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_kotlin--canary.270.3956267287.kt)  
[color_smbusiness_react-native--canary.270.3956267287.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_react-native--canary.270.3956267287.ts)  
[PlasmaTokensColor--canary.270.3956267287.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/PlasmaTokensColor--canary.270.3956267287.swift)  
[typo_mage_ios-swift--canary.270.3956267287.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_ios-swift--canary.270.3956267287.swift)  
[typo_mage_kotlin--canary.270.3956267287.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_kotlin--canary.270.3956267287.kt)  
[typo_mage_react-native--canary.270.3956267287.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_react-native--canary.270.3956267287.ts)  
[typo_plasma_ios-swift--canary.270.3956267287.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_ios-swift--canary.270.3956267287.swift)  
[typo_plasma_kotlin--canary.270.3956267287.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_kotlin--canary.270.3956267287.kt)  
[typo_plasma_react-native--canary.270.3956267287.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_react-native--canary.270.3956267287.ts)  
[typo_ruler_ios-swift--canary.270.3956267287.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_ios-swift--canary.270.3956267287.swift)  
[typo_ruler_kotlin--canary.270.3956267287.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_kotlin--canary.270.3956267287.kt)  
[typo_ruler_react-native--canary.270.3956267287.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_react-native--canary.270.3956267287.ts)  
[typo_sage_ios-swift--canary.270.3956267287.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_ios-swift--canary.270.3956267287.swift)  
[typo_sage_kotlin--canary.270.3956267287.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_kotlin--canary.270.3956267287.kt)  
[typo_sage_react-native--canary.270.3956267287.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_react-native--canary.270.3956267287.ts)  
[typo_sbermarket_ios-swift--canary.270.3956267287.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_ios-swift--canary.270.3956267287.swift)  
[typo_sbermarket_kotlin--canary.270.3956267287.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_kotlin--canary.270.3956267287.kt)  
[typo_sbermarket_react-native--canary.270.3956267287.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_react-native--canary.270.3956267287.ts)  
[typo_soulmate_ios-swift--canary.270.3956267287.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_ios-swift--canary.270.3956267287.swift)  
[typo_soulmate_kotlin--canary.270.3956267287.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_kotlin--canary.270.3956267287.kt)  
[typo_soulmate_react-native--canary.270.3956267287.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_react-native--canary.270.3956267287.ts)

<details>
  <summary>📦 Published PR as canary version: <code>Canary Versions</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @salutejs/plasma-b2c@1.126.0-canary.270.3956267287.0
  npm install @salutejs/plasma-core@1.89.0-canary.270.3956267287.0
  npm install @salutejs/plasma-icons@1.117.0-canary.270.3956267287.0
  npm install @salutejs/plasma-temple@1.123.0-canary.270.3956267287.0
  npm install @salutejs/plasma-ui@1.156.0-canary.270.3956267287.0
  npm install @salutejs/plasma-web@1.153.0-canary.270.3956267287.0
  npm install @salutejs/plasma-cy-utils@0.39.0-canary.270.3956267287.0
  npm install @salutejs/plasma-sb-utils@0.87.0-canary.270.3956267287.0
  # or 
  yarn add @salutejs/plasma-b2c@1.126.0-canary.270.3956267287.0
  yarn add @salutejs/plasma-core@1.89.0-canary.270.3956267287.0
  yarn add @salutejs/plasma-icons@1.117.0-canary.270.3956267287.0
  yarn add @salutejs/plasma-temple@1.123.0-canary.270.3956267287.0
  yarn add @salutejs/plasma-ui@1.156.0-canary.270.3956267287.0
  yarn add @salutejs/plasma-web@1.153.0-canary.270.3956267287.0
  yarn add @salutejs/plasma-cy-utils@0.39.0-canary.270.3956267287.0
  yarn add @salutejs/plasma-sb-utils@0.87.0-canary.270.3956267287.0
  ```
</details>

### CODE REVIEW

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-270/

File: website/plasma-theme-builder/src/api/githubFilesFetcher.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+            });
+
+            return result.data;
+        } catch (error) {
+            // eslint-disable-next-line no-console
+            console.error(error);
+        }
+    };
+
+    const filesSource = await getFileSource(path);
```
Comment (Yeti-or) [inline website/plasma-theme-builder/src/api/githubFilesFetcher.ts:28]: а типов у него нет ?
почему нужен такой гард на isString ?

File: website/plasma-theme-builder/src/components/hooks.tsx [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+            } catch (err) {
+                setErrorMessage(`Тема "${themeName}" не найдена в репозитории.`);
+                console.error(err);
+            }
+        };
+
+        fetchData();
+    }, [themeName]);
+
+    return [response, errorMessage] as const;
```
Comment (Yeti-or) [inline website/plasma-theme-builder/src/components/hooks.tsx:29]: а что такое `as const` ?

File: website/plasma-theme-builder/src/components/Error/Error.tsx [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import { Button } from '@salutejs/plasma-b2c';
+
+const StyledError = styled.div``;
+
+const Description = styled.div`
+    margin-top: 2rem;
+`;
+
+const StyledButton = styled(Button)`
+    display: block;
```
Comment (Yeti-or) [inline website/plasma-theme-builder/src/components/Error/Error.tsx:12]: display block ?

File: website/plasma-theme-builder/src/api/githubFilesFetcher.ts (hunk 1 above)
Comment (neretin-trike) [reply website/plasma-theme-builder/src/api/githubFilesFetcher.ts:28]: да, там тип несколько кривой получается. На самом деле в filesSource, если запрашивать контент через их апи, вернётся строка, а не тот набор полей, который у них указан

File: website/plasma-theme-builder/src/api/githubFilesFetcher.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    return isString(filesSource) ? filesSource : '';
+};
+
+export const getThemeData = async (themeName: string, owner = 'salute-developers', repo = 'plasma', token?: string) =>
```
[the remark this answers is not in the document]
Comment (neretin-trike) [reply website/plasma-theme-builder/src/api/githubFilesFetcher.ts:33]: не, тут токен для запроса в принципе не нужен

А для авторизации через гитхаб достаточно открытого client_id робота на клиенте, который уже будет генерить токен и отдавать пользователю

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🔴 Fail**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-270/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-270/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-270/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-270/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-270/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-270/temple-storybook/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-270/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-270/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-270/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-270/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-270/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-270/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-270/temple-storybook/

File: website/plasma-theme-builder/src/components/Error/Error.tsx (hunk 3 above)
Comment (Yakutoc) [reply website/plasma-theme-builder/src/components/Error/Error.tsx:12]: Ты про то что не стал делать дополнительную обертку с флекс контейнером?

File: website/plasma-theme-builder/src/components/Token/Token.tsx [hunk 5]
```diff
@@ -80,7 +80,7 @@ export const Token = ({ name, data }: TokenProps) => {
     return (
         <StyledToken enabled={enabled}>
             <Tooltip placement="top-start" isVisible={Boolean(comment) && visible} arrow text={comment || ''} animated>
-                <TokenName disable={!enabled} onMouseLeave={onMouseLeave} onMouseEnter={onMouseEnter}>
+                <TokenName disable={enabled === false} onMouseLeave={onMouseLeave} onMouseEnter={onMouseEnter}>
```
Comment (Yakutoc) [inline website/plasma-theme-builder/src/components/Token/Token.tsx:83]: @neretin-trike Уточню, а почему заменил на такую запись?

**Comment by Yakutoc** [comment]
@neretin-trike Опиши коротко пожалуйста, что и как будет работать в этом PR.

File: website/plasma-theme-builder/src/components/Error/Error.tsx (hunk 3 above)
Comment (neretin-trike) [reply website/plasma-theme-builder/src/components/Error/Error.tsx:12]: да

File: website/plasma-theme-builder/src/components/Token/Token.tsx (hunk 5 above)
Comment (neretin-trike) [reply website/plasma-theme-builder/src/components/Token/Token.tsx:83]: там была ошибка. В моём случае, отключенным токен считается только если у него явно проставлено значение false

**Review by Yakutoc — APPROVED** [review]

File: website/plasma-theme-builder/src/components/Error/Error.tsx (hunk 3 above)
Comment (Yeti-or) [reply website/plasma-theme-builder/src/components/Error/Error.tsx:12]: блин я слишком стар для этого
и сделал бы `float: right`

Comment (neretin-trike) [reply website/plasma-theme-builder/src/components/Error/Error.tsx:12]: :D тоже нормас

**Pull request merged**